index.bs

```diff
@@ -357,9 +357,40 @@ is defined by running the following steps:
 
 The <dfn>writeDatagrams</dfn> algorithm is given a |transport| as parameter and
 |data| as input. It is defined by running the following steps:
-1. TODO
-1. Enqueue data for datagram transmission.
-1. Return [=a promise resolved with=] undefined.
+1. Let |timestamp| be a timestamp representing now.
+1. Let |promise| be a new promise.
+1. If |data| is not a {{Uint8Array}} object, then return [=a promise rejected with=] a {{TypeError}}.
+1. Let |bytes| be a copy of bytes which |data| represents.
+1. Let |chunk| be a tuple of |bytes|, |timestamp| and |promise|.
+1. Enqueue |chunk| to |transport|'s {{[[OutgoingDatagramsExpirationDuration]]}}.
+1. If the length of |transport|'s {{[[OutgoingDatagramsQueue]]}} is less than
+   |transport|'s {{[[OutgoingDatagramsHighWaterMark]]}}, then [=resolve=] |promise| with undefined.
+1. Return |promise|.
+
+Note: The associated {{WritableStream}} calls [=writeDatagrams=] only when all the promises that
+have been returned by [=writeDatagrams=] have been resolved. Hence the timestamp and the expiration
+duration work well only when the web developer takes care of
+{{WritableStreamDefaultWriter/ready|WritableStreamDefaultWriter.ready}}.
+
+To <dfn>sendDatagrams</dfn>, given a {{WebTransport}} object |transport|, run these steps:
+1. Let |queue| be |transport|'s {{[[OutgoingDatagramsQueue]]}}.
+1. Let |duration| be |transport|'s {{[[OutgoingDatagramsExpirationDuration]]}}.
+1. If |duration| is null, then set |duration| to an implementation defined {{double}} value.
+1. while |queue| is not empty:
+  1. Let |bytes|, |timestamp| and |promise| be |queue|'s first element.
+  1. If more than |duration| milliseconds have passed since |timestamp|, then:
+     1. [=Resolve=] |promise| with undefined.
+     1. Remove the first element from |queue|.
+  1. Otherwise, break this loop.
+1. while |queue| is not empty:
+  1. Let |bytes|, |timestamp| and |promise| be |queue|'s first element.
+  1. If it is not possible to send |bytes| to the network immediately, then break this loop.
+  1. Send |bytes| to the network. [[!WEB-TRANSPORT-HTTP3]]
+  1. [=Resolve=] |promise| with undefined.
+  1. Remove the first element from |queue|.
+
+The user agent MAY run [=sendDatagrams=] with any {{WebTransport}} object whose
```

Comment:
I interpret this to mean that an implementation which never ran `sendDatagrams` would be valid. Which would be strange. In general we want `sendDatagrams` to be run in a timely fashion. I'm not sure how we can tighten the wording to limit the ability of implementations to be terrible. Maybe something like
```
The user agent SHOULD run [=sendDatagrams=] for any {{WebTransport}} object
whose {{[[WebTransportState]]}} is `"connected"` as soon as reasonably
possible whenever the algorithm can make progress.
```
We maybe need to say what task queue it should run on, I'm not sure.

Reply:
Done. I'm not sure about the task queue too.